keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,167 @@
+See also https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager
+
+## Design Details
+### Feature Gate
+To set `--feature-gates=MemoryQoS=true` to enable the feature.
+
+### Rules
+#### memory.low vs memory.min
+In cgroups v2, `memory.low` is designed for best-effort memory protection which is more like "soft guarantee" and won't be reclaimed unless memory can't be reclaimed from any unprotected cgroups. `memory.min` is a bit aggressive. It will always retain specified amount of memory and it can be never reclaimed. When requirement is not satisfied, system OOM killer will be invoked. 
+
+#### Mapping Details
+1. If container sets `requests.memory`, we set `memory.min=requests.memory` at container level cgroup
+2. If any containers in pod sets `requests.memory`, we set `memory.min=sum(pod.spec.containers.resources.requests[memory])` at pod level cgroup
+3. If kubelet sets `kube-reserved` and `kube-reserved-cgroup`, we set `memory.min=kube-reserved` at `kube-reserved-cgroup` node level cgroup
+4. If kubelet sets `system-reserved` and `system-reserved-cgroup`, we set `memory.min=system-reserved` at `system-reserved-cgroup` node level cgroup
+5. There are no changes regarding memory limit, that is `memory.max=memory limits` (same as existing cgroup v2 implementation)
+
+### Interactive
+New `Unified` field will be added in both CRI and QoS Manager for cgroups v2 extra parameters. It is recommended to has same semantics with opencontainers/runtime-spec#1040
+
+- container level: `Unified` added in `LinuxContainerResources`
+- pod level: `Unified` added in `cm.ResourceConfig` 
+- node level: `Unified` added in `cm.ResourceConfig` 
+
+### Workflow
+#### Container Level Cgroup
+![](./container-level.png)
+
+#### Pod Level Cgroup
+![](./pod-level.png)
+
+#### Node Level Cgroup
+![](./node-level.png)
+
+### Cgroup Hierarchy 
+Pod cgroup:
+```
+/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers.resources.requests[memory])
+```
+
+Container cgroup:
+```
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
+```
+
+Node cgroup:
+```
+/cgroup2/<kube-reserved-cgroup|system-reserved-cgroup>/memory.min=<kube-reserved|system-reserved-cgroup>
+```
+
+### Cgroup v2 Support
+ After Kubernetes v1.19, kubelet can identify cgroups v2 and do the convention. Since [v1.0.0-rc93](https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc93), runc supports `Unified` to pass through cgroups v2 parameters. So we use this variable to pass `memory.min` when cgroups v2 mode is detected.
+
+### Container Runtime Interface (CRI) Changes
+We should add a new field `Unified` in CRI api which is basically passthrough for OCI spec Unified field and has same semantics: opencontainers/runtime-spec#1040
+
+```
+type LinuxContainerResources struct {
+    ...
+    Unified    []string `protobuf:"bytes,1,opt,name=unified,proto3" json:"unified,omitempty"`
```

Comment:
Also, what would be the semantics of these? For OCI the container runtime should create an error if it cannot enforce these values, and we should probably do the same for these, or? Imo. we should at least make that very clear before creating adding this.

Ie. if a container runtime doesn't support `memory.min`, should there be a kubelet flag to disable it? If node does not support it, it is fairly easy to just don't send it to CRI. But if the node supports it but not the container runtime, what to do?

Reply:
> Shouldn't this be something like `Map<String, String>` instead?
> 
> https://github.com/opencontainers/runtime-spec/blob/7c549cb0939af03d5a2a8b271e2ad6871309e228/specs-go/config.go#L376

It's a typo, I have updated.

> Also, what would be the semantics of these? For OCI the container runtime should create an error if it cannot enforce these values, and we should probably do the same for these, or? Imo. we should at least make that very clear before creating adding this.

I think OCI should return error for unsupported values. It may need to be considered in conjunction with cgroups v2 supports.

> Ie. if a container runtime doesn't support `memory.min`, should there be a kubelet flag to disable it? If node does not support it, it is fairly easy to just don't send it to CRI. But if the node supports it but not the container runtime, what to do?
I have added a kubelet flag `--enforce-memory-qos=<container, pod, qos, node>` to enable it at different levels.